### PR TITLE
added reshape transpose matmul fuse pass

### PR DESCRIPTION
**Pull request by sfraczek**
Introduce MKLDNN fuse pass (designed especially for Ernie) named `reshape_transpose_matmul_mkldnn_fuse_pass`. It fusses reshape and transpose operations in the inputs of matmul into the matmul operator itself.

To acheive this it adds 4 new attributes to matmul_op.
* fused_transpose_X, 
* fused_transpose_Y,
* fused_reshape_X, 
* fused_reshape_Y.

### CODE REVIEW

**Comment by sfraczek** [comment]
Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz | % improvement
-- | --
ERNIE_fp32_performance | 4,41%
ERNIE_fp32_20_threads_performance | 15,08%
ERNIE_int8_performance | 6,12%
ERNIE_int8_20_threads_performance | 16,47%
TA_B_bert_performance | -0,51%



**updated** 17 IV 2020 14:44

**Comment by sfraczek** [comment]
This PR is related to another PR https://github.com/PaddlePaddle/Paddle/pull/23866

File: paddle/fluid/framework/ir/mkldnn/reshape_transpose_matmul_fuse_pass_tester.cc
Comment (lidanqing-vv) [inline paddle/fluid/framework/ir/mkldnn/reshape_transpose_matmul_fuse_pass_tester.cc]: PADDLE_ENFORCE_NOT_NULL(shape_correct);

File: paddle/fluid/framework/ddim.h
Comment (lidanqing-vv) [inline paddle/fluid/framework/ddim.h]: 不允许含有&&，即对子表达式需要分别进行报错
&& is not allowed in PADDLE_ENFORCE_*, you can write it separately.

File: paddle/fluid/operators/matmul_op.cc
Comment (lidanqing-vv) [inline paddle/fluid/operators/matmul_op.cc]: This one is good, but maybe you need to ask luotao for approval(not sure, let's wait for CI report). In new regulation, || is allowed for PADDLE_ENFORCE should not report error.

File: paddle/fluid/framework/ir/mkldnn/reshape_transpose_matmul_fuse_pass_tester.cc
Comment (lidanqing-vv) [inline paddle/fluid/framework/ir/mkldnn/reshape_transpose_matmul_fuse_pass_tester.cc]: PADDLE_ENFORCE_NOT_NULL

**Review by lidanqing-vv — COMMENTED** [review]
Please refer to [PADDLE_ENFORCE-Rewriting-Specification](https://github.com/PaddlePaddle/Paddle/wiki/PADDLE_ENFORCE-Rewriting-Specification) modify the PR.

File: paddle/fluid/framework/ir/mkldnn/reshape_transpose_matmul_fuse_pass_tester.cc
Comment (sfraczek) [reply paddle/fluid/framework/ir/mkldnn/reshape_transpose_matmul_fuse_pass_tester.cc]: ok I changed it to something different please see new version when I push it.

File: paddle/fluid/framework/ddim.h
Comment (sfraczek) [reply paddle/fluid/framework/ddim.h]: Hmmm, but I copied it from here: https://github.com/PaddlePaddle/Paddle/blob/d33c4343e113d06bb6583b79d135740416dca502/paddle/fluid/operators/transpose_op.cc#L50-L51
Should I still change it?

File: paddle/fluid/framework/ir/mkldnn/reshape_transpose_matmul_fuse_pass_tester.cc
Comment (sfraczek) [reply paddle/fluid/framework/ir/mkldnn/reshape_transpose_matmul_fuse_pass_tester.cc]: ok I changed it to something different please see new version when I push it.

File: paddle/fluid/framework/ddim.h
Comment (lidanqing-vv) [reply paddle/fluid/framework/ddim.h]: Hi, @sfraczek  Yes, all the files you change in this PR need to update the PADDLE_ENFORCE. Because the CI will scan your PR files. Not only places you modified, but also those places that passed before, because you change the file, they will not pass this time.
Soon we need to upgrade PADDLE_ENFORCE in all mkl-dnn ops, because others like CPU native ops files have all upgraded. I will make a jira ticket for this.

File: paddle/fluid/framework/ir/graph_pattern_detector.cc
Comment (lidanqing-vv) [inline paddle/fluid/framework/ir/graph_pattern_detector.cc]: You did not add any log in this file~

File: paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc
Comment (lidanqing-vv) [inline paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc:110]: You don't need to use for loop,  STL will be faster and shorter too
you can include `<algorithm.h>` header
then
```
bool all_less_than_size= std::all_of(axis.begin(),axis.end(), [](int i){ return i<axis.size(); });
PADDLE_ENFORCE_EQ(all_less_than_size, true, "value must be ranging from 0 to (dims - 1), but....");
PADDLE_ENFORCE_EQ(std::unique(axis.begin(), axis.end()), axis.end(),"In axis array, elements must be unique, but....");
```

File: paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc
Comment (lidanqing-vv) [inline paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc]: If it only for size checking, no need to declare a new variable, I think

File: paddle/fluid/framework/ddim.h
Comment (lidanqing-vv) [inline paddle/fluid/framework/ddim.h]: You may consider change this as another place comment

File: paddle/fluid/framework/ir/graph_pattern_detector.cc
Comment (lidanqing-vv) [inline paddle/fluid/framework/ir/graph_pattern_detector.cc]: This page did not use prettylog

File: paddle/fluid/operators/matmul_op.cc
Comment (lidanqing-vv) [inline paddle/fluid/operators/matmul_op.cc]: I am worried this will cause CI report error, because no description. I know it is pain, I also need to watch out when making a PR :(

File: python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py
Comment (lidanqing-vv) [inline python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py:182]: Do we need to check if the machine support avx and then set it, if not support give some information to user.

File: python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py
Comment (lidanqing-vv) [inline python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py:174]: Why the Y input of matmul has to be retransposed again?

File: python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py
Comment (lidanqing-vv) [inline python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py:214]: Because self.x and self.y will be assigned as input of matmul mkldnn op, they should be simple as we only check if matmul mkldnn op working. Do you think self.x and self.y could be simplified? 
when we calculate self.out, we can do a lot of transpose and reshape etc.

**Comment by lidanqing-vv** [comment]
I have one doubt. Your and Artur's PR both did not change the key after adding attributes, but I assume the matmul op with `reshapeX` set should be different than ops with `reshapeX` etc set. If you don't add attributes to the key, it could cause cache mechanism problems, by which I mean in next iteration reusing the op. But why the whole inference in this case achieve good accuracy?

File: paddle/fluid/framework/ddim.h
Comment (wojtuss) [inline paddle/fluid/framework/ddim.h]: You can unify and skip `framework::` here.

File: paddle/fluid/framework/ddim.h
Comment (wojtuss) [inline paddle/fluid/framework/ddim.h]: What does the `x_` stand for? Maybe `in_rank` would be good enough?

**Comment by sfraczek** [comment]
Maybe output_name uniquely identifies matmul in network? But in unit tests it can fail maybe I'm not sure. But why we do not reuse primitive if it is the same but has different output_name?

File: python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py
Comment (sfraczek) [reply python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py:182]: I copied this from Michal's test to be safe https://github.com/PaddlePaddle/Paddle/pull/23754/files/eda126700bd34bd0650ff57999ccc4fd04a4d5de#diff-68b25bf1ad071a1bf85630c689dd071bR36

File: paddle/fluid/framework/ddim.h
Comment (sfraczek) [reply paddle/fluid/framework/ddim.h]: ok i changed it

File: paddle/fluid/operators/matmul_op.cc
Comment (sfraczek) [reply paddle/fluid/operators/matmul_op.cc]: ok

File: paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc
Comment (sfraczek) [reply paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc]: i would leave it. it was the same in transpose_op.cc

File: python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py
Comment (sfraczek) [reply python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py:174]: transpose_y is always true

File: paddle/fluid/framework/ir/graph_pattern_detector.cc
Comment (sfraczek) [reply paddle/fluid/framework/ir/graph_pattern_detector.cc]: ok

File: paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc
Comment (sfraczek) [reply paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc:110]: ok

File: paddle/fluid/framework/ir/graph_pattern_detector.cc
Comment (sfraczek) [reply paddle/fluid/framework/ir/graph_pattern_detector.cc]: ok

File: python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py
Comment (sfraczek) [reply python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py:214]: it is as I explained on the call i need to compare reshape+transpose+matmul with numpy matmul

File: paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc
Comment (sfraczek) [reply paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc:110]: this doesnt work because it modifies axis

**Comment by sfraczek** [comment]
Please check new version

**Comment by lidanqing-vv** [comment]
@sfraczek  How is the CIs failures, are they random failures?

**Comment by sfraczek** [comment]
PR_CI_CPU_Py2 now only complains about `@skip_check_grad_ci`
```
[09:19:32] :	 [Step 1/1] 0. It is an Op accuracy problem, please take care of it. You must have one RD (zhangting2020 (Recommend), luotao1 or phlrain) approval for the usage (either add or delete) of @skip_check_grad_ci. For more information, please refer to: https://github.com/PaddlePaddle/Paddle/wiki/Gradient-Check-Is-Required-for-Op-Test. The corresponding lines are as follows:
[09:19:32] :	 [Step 1/1] 
[09:19:32] :	 [Step 1/1] python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py
[09:19:32] :	 [Step 1/1] 
[09:19:32] :	 [Step 1/1] +	@skip_check_grad_ci(reason="DNNL's MatMul doesn't implement grad kernel.")
[09:19:32] :	 [Step 1/1] 
[09:19:32] :	 [Step 1/1] 
[09:19:32] :	 [Step 1/1] There are 1 approved errors.
```

**Comment by lidanqing-vv** [comment]
> PR_CI_CPU_Py2 now only complains about `@skip_check_grad_ci`
> 
> ```
> [09:19:32] :	 [Step 1/1] 0. It is an Op accuracy problem, please take care of it. You must have one RD (zhangting2020 (Recommend), luotao1 or phlrain) approval for the usage (either add or delete) of @skip_check_grad_ci. For more information, please refer to: https://github.com/PaddlePaddle/Paddle/wiki/Gradient-Check-Is-Required-for-Op-Test. The corresponding lines are as follows:
> [09:19:32] :	 [Step 1/1] 
> [09:19:32] :	 [Step 1/1] python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py
> [09:19:32] :	 [Step 1/1] 
> [09:19:32] :	 [Step 1/1] +	@skip_check_grad_ci(reason="DNNL's MatMul doesn't implement grad kernel.")
> [09:19:32] :	 [Step 1/1] 
> [09:19:32] :	 [Step 1/1] 
> [09:19:32] :	 [Step 1/1] There are 1 approved errors.
> ```

When all other CIs passed, Luotao will approve this CI

File: paddle/fluid/framework/ddim.h
Comment (wojtuss) [inline paddle/fluid/framework/ddim.h]: But -> but

File: paddle/fluid/framework/ddim.h
Comment (wojtuss) [inline paddle/fluid/framework/ddim.h]: Maybe: "... Each element of the axis must appear exactly once in the range from 0 to (dims - 1), where ..., but ..."

File: paddle/fluid/framework/ir/mkldnn/reshape_transpose_matmul_fuse_pass.h
Comment (wojtuss) [inline paddle/fluid/framework/ir/mkldnn/reshape_transpose_matmul_fuse_pass.h]: I think it would be good if the name of the pass reflected that it is MKL-DNN pass.

**Comment by sfraczek** [comment]
PR-CI-Coverage failed with unknown to me reason - it works on my desktop and on clx server:
```
2020-04-17 20:10:50   50/447 Test #102: test_reshape_transpose_matmul_fuse_pass .....................Child aborted***Exception:   0.20 sec
2020-04-17 20:10:50  [==========] Running 2 tests from 1 test case.
2020-04-17 20:10:50  [----------] Global test environment set-up.
2020-04-17 20:10:50  [----------] 2 tests from ReshapeTransposeMatmulFusePass
2020-04-17 20:10:50  [ RUN      ] ReshapeTransposeMatmulFusePass.both_matmul_inputs_reshape_transpose
2020-04-17 20:10:50  WARNING: Logging before InitGoogleLogging() is written to STDERR
2020-04-17 20:10:50  I0417 12:10:50.631882 21783 graph_pattern_detector.cc:100] ---  detected 2 subgraphs
2020-04-17 20:10:50  [37m---    Fused 2 ReshapeTransposeMatmul patterns[0m
2020-04-17 20:10:50  [37m---    Fused 0 ReshapeTransposeMatmul patterns with transpose's xshape[0m
2020-04-17 20:10:50  [37m---    Fused 0 ReshapeTransposeMatmul patterns with reshape's xshape[0m
2020-04-17 20:10:50  [37m---    Fused 0 ReshapeTransposeMatmul patterns with reshape's xshape with transpose's xshape[0m
2020-04-17 20:10:50 test_reshape_transpose_matmul_fuse_pass: ../nptl/pthread_mutex_lock.c:81: __pthread_mutex_lock: Assertion `mutex->__data.__owner == 0' failed.
2020-04-17 20:10:50  
```

File: python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py
Comment (wojtuss) [inline python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py:210]: This and the next test differ only in the datatype. The datatype could be added as a parameter, without duplicating the code. A similar situation is with the other tests below.

**Review by wojtuss — COMMENTED** [review]
I found no major issues, only minor or aesthetic points, which I do not consider as blocking issues.

**Comment by luotao1** [comment]
PR_CI_Coverage is 6148 with 24 cores and 2 V100 GPU cards. 
> Child aborted

We met a similar error in `pyramid_hash` op, see #20698. There are a lot of `Child aborted` in the unit-test, and we could not reproduce on the local machine and don't know why.

The previous log is [Paddle_PR_CI_Coverage_47998.log.zip](https://github.com/PaddlePaddle/Paddle/files/4493615/Paddle_PR_CI_Coverage_47998.log.zip)

Since this op is in `paddle.fluid.contrib` and useless, we use a trick method...., i.e, we don't test this Op on PR_CI_Coverage.

https://github.com/PaddlePaddle/Paddle/blob/fb98116cd7ceb22021a3fd2c2cf9bdf3d1d3708d/paddle/fluid/operators/CMakeLists.txt#L57-L60

File: paddle/fluid/framework/ddim.h
Comment (sfraczek) [reply paddle/fluid/framework/ddim.h]: ok

File: paddle/fluid/framework/ir/mkldnn/reshape_transpose_matmul_fuse_pass.h
Comment (sfraczek) [reply paddle/fluid/framework/ir/mkldnn/reshape_transpose_matmul_fuse_pass.h]: Where should it be added?

File: python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py
Comment (sfraczek) [reply python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py:210]: I thought about it but I don't know a good way to do this

Comment (wojtuss) [reply python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py:210]: You could define a method
```
def init_data_type(self):
   self.data_type_ = 'float32'
```
then call it in the `setUp()` method, and use `self.data_type_` in the body of the `generate_data()` method instead of using the `"float32"` string. In the next class only override the `init_data_type()` method for INT8 test.

File: paddle/fluid/framework/ir/mkldnn/reshape_transpose_matmul_fuse_pass.h
Comment (wojtuss) [reply paddle/fluid/framework/ir/mkldnn/reshape_transpose_matmul_fuse_pass.h]: Maybe `reshape_transpose_matmul_mkldnn_fuse_pass`?

**Comment by sfraczek** [comment]
@luotao1  
This `if(WITH_COVERAGE OR NOT WITH_AVX OR WIN32) ` looks to me like it runs on coverage because of WITH_COVERAGE? Is this really avoid building on coverage?

File: python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py
Comment (sfraczek) [reply python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py:210]: ok thanks!

File: paddle/fluid/framework/ir/mkldnn/reshape_transpose_matmul_fuse_pass.h
Comment (sfraczek) [reply paddle/fluid/framework/ir/mkldnn/reshape_transpose_matmul_fuse_pass.h:26]: should it also be `ReshapeTransposeMatmulMkldnnPattern`?

**Comment by luotao1** [comment]
Sorry, I refer to a wrong code, please see
https://github.com/PaddlePaddle/Paddle/blob/fb98116cd7ceb22021a3fd2c2cf9bdf3d1d3708d/python/paddle/fluid/tests/unittests/CMakeLists.txt#L86-L89

File: paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc
Comment (Sand3r-) [reply paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc:110]: Danqing had a good point, however `std::unique` indeed modifies the container. Alternatively, you could create a set from vector, and see if the sizes are the same. The first Enforce should work just fine.
```C++
PADDLE_ENFORCE_EQ(std::set(axis.begin(), axis.end()).size(), axis.size(), "In an axis array, elements must be unique, however...");
```
It's all just syntactic sugar anyway.

File: paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc
Comment (Sand3r-) [inline paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc]: You may want to save `batch_size_` to a local variable to avoid unnecessary cache misses due to aliasing problem.
Similarly, you may want to compute the `<smth>_offset_`s here instead of storing them as class members.

**Review by Sand3r- — COMMENTED** [review]
Sylwek, thank you so much for that rich contribution and hard work you've put into this fuse. Please refer to my comments, **and** to [my review in Artur's PR](https://github.com/PaddlePaddle/Paddle/pull/23866#pullrequestreview-397080637). All comments given there also apply here.

**Comment by lidanqing-vv** [comment]
@sfraczek  PR-CI-Coverage failed
```
2020-04-22 04:40:55   51/454 Test #103: test_reshape_transpose_matmul_mkldnn_fuse_pass .....Child aborted***Exception:   0.37 sec
2020-04-22 04:40:55  [==========] Running 2 tests from 1 test case.
2020-04-22 04:40:55  [----------] Global test environment set-up.
2020-04-22 04:40:55  [----------] 2 tests from ReshapeTransposeMatmulMkldnnFusePass
2020-04-22 04:40:55  [ RUN      ] ReshapeTransposeMatmulMkldnnFusePass.both_matmul_inputs_reshape_transpose
2020-04-22 04:40:55  WARNING: Logging before InitGoogleLogging() is written to STDERR
2020-04-22 04:40:55  I0421 20:40:55.324996 26349 graph_pattern_detector.cc:100] ---  detected 2 subgraphs
2020-04-22 04:40:55  [37m---    Fused 2 ReshapeTransposeMatmulMkldnn patterns[0m
2020-04-22 04:40:55  [37m---    Fused 0 ReshapeTransposeMatmulMkldnn patterns with transpose's xshape[0m
2020-04-22 04:40:55  [37m---    Fused 0 ReshapeTransposeMatmulMkldnn patterns with reshape's xshape[0m
2020-04-22 04:40:55  [37m---    Fused 0 ReshapeTransposeMatmulMkldnn patterns with reshape's xshape with transpose's xshape[0m
2020-04-22 04:40:55  test_reshape_transpose_matmul_mkldnn_fuse_pass: ../nptl/pthread_mutex_lock.c:81: __pthread_mutex_lock: Assertion `mutex->__data.__owner == 0' failed.
```

Sylwek, did you have this error on your own machine? If you dont have this error on your own machine, you may try to reproduce it on 6148, 6248 or 6271.

**Comment by lidanqing-vv** [comment]
@luotao1  On which machine 效率云 is running these tests? Is that 6148 physical machine? We always have problem that locally passed, but CI failed. We would like to reproduce. Thank you!

**Comment by luotao1** [comment]
> Is that 6148 physical machine

6148 with 24 cores, not a physical machine, it's on cloud

**Comment by sfraczek** [comment]
> @sfraczek PR-CI-Coverage failed
> 
> ```
> 2020-04-22 04:40:55   51/454 Test #103: test_reshape_transpose_matmul_mkldnn_fuse_pass .....Child aborted***Exception:   0.37 sec
> 2020-04-22 04:40:55  [==========] Running 2 tests from 1 test case.
> 2020-04-22 04:40:55  [----------] Global test environment set-up.
> 2020-04-22 04:40:55  [----------] 2 tests from ReshapeTransposeMatmulMkldnnFusePass
> 2020-04-22 04:40:55  [ RUN      ] ReshapeTransposeMatmulMkldnnFusePass.both_matmul_inputs_reshape_transpose
> 2020-04-22 04:40:55  WARNING: Logging before InitGoogleLogging() is written to STDERR
> 2020-04-22 04:40:55  I0421 20:40:55.324996 26349 graph_pattern_detector.cc:100] ---  detected 2 subgraphs
> 2020-04-22 04:40:55  �[37m---    Fused 2 ReshapeTransposeMatmulMkldnn patterns�[0m
> 2020-04-22 04:40:55  �[37m---    Fused 0 ReshapeTransposeMatmulMkldnn patterns with transpose's xshape�[0m
> 2020-04-22 04:40:55  �[37m---    Fused 0 ReshapeTransposeMatmulMkldnn patterns with reshape's xshape�[0m
> 2020-04-22 04:40:55  �[37m---    Fused 0 ReshapeTransposeMatmulMkldnn patterns with reshape's xshape with transpose's xshape�[0m
> 2020-04-22 04:40:55  test_reshape_transpose_matmul_mkldnn_fuse_pass: ../nptl/pthread_mutex_lock.c:81: __pthread_mutex_lock: Assertion `mutex->__data.__owner == 0' failed.
> ```
> 
> Sylwek, did you have this error on your own machine? If you dont have this error on your own machine, you may try to reproduce it on 6148, 6248 or 6271.

it passes on 6148,6248.

**Review by wojtuss — APPROVED** [review]
LGTM

**Comment by sfraczek** [comment]
@luotao1 please see

**Review by Sand3r- — APPROVED** [review]

**Comment by lidanqing-vv** [comment]
Hi, @luotao1  Could you please review this PR? If not, could you please merge it to develop? This will not be in release1.8 because now QAT team already started verifying release/2.0-beta

File: paddle/fluid/operators/matmul_op.cc
Comment (luotao1) [inline paddle/fluid/operators/matmul_op.cc:526]: Are there too many attributes?

Comment (wojtuss) [reply paddle/fluid/operators/matmul_op.cc:526]: @luotao1 , these attributes are necessary to keep information from `reshape2+transpose2+matmul` fuse.

**Review by luotao1 — APPROVED** [review]
LGTM, please supply your PR description

**Comment by sfraczek** [comment]
@luotao1 ok, I added PR description.

**Pull request merged**
